rs/crypto/internal/crypto_service_provider/src/vault/local_csp_vault/vetkd/mod.rs

```diff
@@ -0,0 +1,115 @@
+use crate::key_id::KeyId;
+use crate::public_key_store::PublicKeyStore;
+use crate::secret_key_store::SecretKeyStore;
+use crate::types::CspSecretKey;
+use crate::vault::api::{VetKdCspVault, VetKdEncryptedKeyShareCreationVaultError};
+use crate::vault::local_csp_vault::LocalCspVault;
+use ic_crypto_internal_bls12_381_vetkd::{
+    DerivationPath, EncryptedKeyShare, G2Affine, PairingInvalidPoint, Scalar, TransportPublicKey,
+    TransportPublicKeyDeserializationError,
+};
+use ic_crypto_internal_logmon::metrics::{MetricsDomain, MetricsResult, MetricsScope};
+use ic_types::crypto::vetkd::VetKdEncryptedKeyShareContent;
+use ic_types::crypto::ExtendedDerivationPath;
+use rand::{CryptoRng, Rng};
+
+#[cfg(test)]
+mod tests;
+
+impl<R: Rng + CryptoRng, S: SecretKeyStore, C: SecretKeyStore, P: PublicKeyStore> VetKdCspVault
+    for LocalCspVault<R, S, C, P>
+{
+    /// Generates an encrypted vetKD key share.
+    fn create_encrypted_vetkd_key_share(
+        &self,
+        key_id: KeyId,
+        master_public_key: Vec<u8>,
+        encryption_public_key: Vec<u8>,
+        derivation_path: ExtendedDerivationPath,
+        derivation_id: Vec<u8>,
+    ) -> Result<VetKdEncryptedKeyShareContent, VetKdEncryptedKeyShareCreationVaultError> {
+        let start_time = self.metrics.now();
+        let result = self.create_encrypted_vetkd_key_share_internal(
+            key_id,
+            master_public_key,
+            encryption_public_key,
+            derivation_path,
+            derivation_id,
+        );
+        self.metrics.observe_duration_seconds(
+            MetricsDomain::VetKd,
+            MetricsScope::Local,
+            "create_encrypted_vetkd_key_share",
+            MetricsResult::from(&result),
+            start_time,
+        );
+        result
+    }
+}
+
+impl<R: Rng + CryptoRng, S: SecretKeyStore, C: SecretKeyStore, P: PublicKeyStore>
+    LocalCspVault<R, S, C, P>
+{
+    fn create_encrypted_vetkd_key_share_internal(
+        &self,
+        key_id: KeyId,
+        master_public_key: Vec<u8>,
+        encryption_public_key: Vec<u8>,
+        derivation_path: ExtendedDerivationPath,
+        derivation_id: Vec<u8>,
+    ) -> Result<VetKdEncryptedKeyShareContent, VetKdEncryptedKeyShareCreationVaultError> {
+        let master_public_key =
+            G2Affine::deserialize(&master_public_key).map_err(|_: PairingInvalidPoint| {
+                VetKdEncryptedKeyShareCreationVaultError::InvalidArgument(format!(
+                    "invalid master public key: 0x{}",
+                    hex::encode(&master_public_key)
+                ))
+            })?;
+
+        let transport_public_key = TransportPublicKey::deserialize(&encryption_public_key)
+            .map_err(|e| match e {
+                TransportPublicKeyDeserializationError::InvalidPublicKey => {
+                    VetKdEncryptedKeyShareCreationVaultError::InvalidArgument(format!(
+                        "invalid encryption public key: 0x{}",
```

Comment:
Do IUC that the size of `encryption_public_key` is only bound by the ingress message size? Should we limit the number of bytes we're displaying in that case? Otherwise someone could be sending large inputs to make us store large hex strings in logs.

Reply:
> Do IUC that the size of encryption_public_key is only bound by the ingress message size?

Good point/question! Indeed, currently this is only bound by the ingress message size. But given that we know that the encryption public key must exactly be 48 bytes (a compressed BLS12-381 G1 element), it would make sense to restrict the size of the encryption public key to 48 bytes, rather than accepting any size, which would also mean we can rely on that fact here in the crypto component (vault) so we don't need to take any additional actions/precautions. WDYT?

I created CRP-2662 for this. PTAL

Reply:
Sounds good! If we ensure the size statically (`[u8; 48]`), then I guess no additional checks would be needed.